sw/device/lib/testing/usb_logging.c

```diff
@@ -0,0 +1,455 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/testing/usb_logging.h"
+
+#include <ctype.h>
+
+#include "sw/device/lib/runtime/print.h"
+#include "sw/device/lib/testing/pinmux_testutils.h"
+#include "sw/device/lib/testing/usb_testutils.h"
+#include "sw/device/lib/testing/usb_testutils_controlep.h"
+
+#define MODULE_ID MAKE_MODULE_ID('u', 's', 'l')
+
+// Size of each logging buffer; data is collected in a buffer until the buffer
+// is full (or a timeout occurs) and is then transmitted as a single transfer
+// at higher throughput.
+#define USBLOG_BUF_SIZE 0x1000u
+
+// Simple double-buffering scheme allows us to achieve higher throughput via
+// the existing usb_testutils layer and usbdev.
+#define USBLOG_NUM_BUFFERS 2u
+
+// Maximum length of the configuration descriptor.
+#define CFG_DSCR_LEN_MAX \
+  (USB_CFG_DSCR_LEN +    \
+   USBLOG_MAX_STREAMS * (USB_INTERFACE_DSCR_LEN + 1u * USB_EP_DSCR_LEN))
+
+typedef struct usb_logging_ctx usb_logging_ctx_t;
+
+typedef struct usb_logging_stream {
+  /**
+   * Context pointer; for use within callback handler.
+   */
+  usb_logging_ctx_t *ctx;
+  /**
+   * Reliable delivery required?
+   */
+  bool reliable;
+  /**
+   * Remapping for non-printable characters required?
+   */
+  bool remap;
+  /**
+   * Counter used to ensure logging data is periodically flushed to the USB.
+   */
+  uint8_t flush_cnt;
+  /**
+   * The USB device endpoint that we're using.
+   */
+  uint8_t ep;
+  /**
+   * Index at which writing into the current buffer shall next occur.
+   */
+  uint16_t wr_offset;
+  /**
+   * Current write buffer.
+   */
+  uint8_t wr_buf;
+  /**
+   * Current send buffer.
+   */
+  uint8_t send_buf;
+  /**
+   * Currently sending a buffer?
+   */
+  bool sending;
+  /**
+   * Data buffers.
+   *
+   * TODO: share a specific number of buffers among the configured number of
+   * logging streams? This could allow a better trade off between stream count
+   * and memory usage.
+   */
+  uint8_t buf[USBLOG_NUM_BUFFERS][USBLOG_BUF_SIZE];
+} usb_logging_stream_t;
+
+struct usb_logging_ctx {
+  /**
+   * Access to usb_testutils layer.
+   */
+  usb_testutils_ctx_t *usbutils;
+  /**
+   * Was the logging responsible for initializing the testutils layer?
+   */
+  bool inited;
+  /**
+   * Number of logging streams.
+   */
+  uint8_t nstreams;
+  /**
+   * Per-stream state information.
+   */
+  usb_logging_stream_t streams[USBLOG_MAX_STREAMS];
+};
+
+/**
+ * USB device context types.
+ */
+static usb_testutils_ctx_t usbdev;
+static usb_testutils_controlep_ctx_t usbdev_control;
+
+// Note: for now there is a single logging context.
+static usb_logging_ctx_t the_ctx;
+
+/**
+ * Pinmux handle
+ */
+static dif_pinmux_t pinmux;
+
+/**
+ * Internal development logging; be careful about using this if stdout has been
+ * redirected!
+ */
+static const bool kDevLogging = false;
+
+/**
+ * Character map, used to ensure that non-printable characters are dropped.
+ */
+static uint8_t map[0x100u];
```

Comment:
Ideally this would be a `const` array, so that it can live in flash. I know that makes initialization less convenient.

Reply:
I'll add some explanatory comments about tx_done and tx_flush; these are callbacks from the software handling the USB device, not functions that are visible externally to code that is using the USB logging.
It's not throttling flush operations, but rather implementing a crude timeout (~32 x 16ms) to ensure that logging data does not linger indefinitely in this software layer when no more log output is being generated.

Reply:
> Ideally this would be a `const` array, so that it can live in flash. I know that makes initialization less convenient.

Well, since we don't have an implementation of isprint() and the code is rather forced to assume ASCII, I've just re-expressed it as a bitmap. That makes it easy and compact to express as a const array definition in the source, although it does also reduce the benefit of making it const :)